### PR TITLE
Add bucketize sampler and dataset for HuBERT Base model training pipeline

### PR DESCRIPTION
**Pull request by nateanl**
The PR adds PyTorch Lightning based training script for HuBERT Base model. There are two iterations of pre-training and 1 iteration of ASR fine-tuning on LibriSpeech dataset.

### CODE REVIEW

File: examples/hubert/dataset/dataset.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        self.exp_dir = Path(exp_dir)
+        tsv_dir = self.exp_dir / "tsv"
+        label_dir = self.exp_dir / "label"
+        f_list, ind_list, len_list = self._get_lists(
+            tsv_dir,
+            dataset,
+            phase_type,
+            min_sample,
+            max_sample
+        )
```
Comment (ejguan) [inline examples/hubert/dataset/dataset.py:127]: I am wondering about the size of these returned lists? Just want to give a heads up that multiprocessing with python list would cause N times memory usage. N represents the number of workers.

Using NumPy array or Tensor would help sharing data across processes without copy.

File: examples/hubert/dataset/dataset.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        self.f_list, self.ind_list, self.len_list = f_list, ind_list, len_list
+        self.labels = self._load_label(label_dir, dataset, phase_type)
```
Comment (ejguan) [inline examples/hubert/dataset/dataset.py:129]: Ditto.

File: examples/hubert/dataset/dataset.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        waveform, length, label = sample
+        waveforms.append(waveform)
+        lengths.append(length)
+        labels.append(label)
+
+    data = torch.zeros(len(batch), max(lengths))
+    for i in range(len(waveforms)):
+        data[i][0:waveforms[i].shape[1]] = waveforms[i][0]
+    lengths = torch.Tensor(lengths)
+    labels = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True)
```
Comment (ejguan) [inline examples/hubert/dataset/dataset.py:213]: Would this op be stateless? I mean we would not do backpropogation here.

File: examples/hubert/dataset/dataset.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        max_token_count: Optional[int] = None,
+        batch_size: Optional[int] = None
+    ) -> None:
+        if max_token_count is not None and batch_size is not None:
+            raise AssertionError(
+                "The ``max_token_count`` and ``batch_size`` can't be both set."
+            )
+        self.data_source = data_source
+        self.max_token_count = max_token_count
+        self.buckets = self._get_buckets(self.data_source, num_buckets)
```
Comment (ejguan) [inline examples/hubert/dataset/dataset.py:44]: Same comment on List with multiprocessing.

File: examples/hubert/dataset/dataset.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    Union,
+)
+from torch import Tensor
+import random
+import torch
+import torchaudio
+from torch.utils.data import Dataset, Sampler
+
+
+class BucketizeSampler(Sampler):
```
Comment (ejguan) [inline examples/hubert/dataset/dataset.py:16]: Even though this would not cause any problems, do you want to make `BucketizeSampler` inheriting `BatchSampler`?

**Review by ejguan — COMMENTED** [review]
In general, looks good to me. I have a few comments/suggestions below.

File: examples/hubert/dataset/dataset.py (hunk 1 above)
Comment (nateanl) [reply examples/hubert/dataset/dataset.py:127]: Good point. The size is actually quite large here. I will refactor it by using Tensor.

File: examples/hubert/dataset/dataset.py (hunk 3 above)
Comment (nateanl) [reply examples/hubert/dataset/dataset.py:213]: Could you elaborate on the stateless? The intention here is to compute the NCE loss batch-wise and apply masking afterwards.

File: examples/hubert/dataset/dataset.py (hunk 5 above)
Comment (ejguan) [reply examples/hubert/dataset/dataset.py:16]: Emm. The __init__ is also varied between `BucketizeSampler` and `Sampler` right? It will give users hints about not using it as a sampler as the argument of DataLoader.

Comment (nateanl) [reply examples/hubert/dataset/dataset.py:16]: I see. I have verified that changing `Sampler` to `BatchSampler` doesn't affect the running. Also I don't need to set the batch_size in the dataloader.

File: examples/hubert/dataset/hubert_dataset.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        self.labels = self._load_labels(label_dir, dataset, phase_type)
+
+    def __len__(self):
+        return len(self.f_list)
+
+    def _get_lists(
+        self,
+        tsv_dir: Path,
+        dataset: str,
+        phase_type: str,
```
Comment (carolineechen) [inline examples/hubert/dataset/hubert_dataset.py:155]: would `subset` be a better name for this?

Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:155]: Sounds good. Let me replace with it.

File: examples/hubert/dataset/hubert_dataset.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+            buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+        return buckets
+
+    def __iter__(self) -> Iterator[List[int]]:
+        iter_list = []
+        total_len = 0
+        batch = []
+        len_list = self.data_source.len_list
+        if self.max_token_count:
+            for k in self.buckets.keys():
```
Comment (ejguan) [inline examples/hubert/dataset/hubert_dataset.py:89]: Do you also want to do a permutation of keys here to randomize the bucket?

File: examples/hubert/dataset/hubert_dataset.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+                    index = self.buckets[k][i]
+                    if total_len + len_list[index] > self.max_token_count:
+                        iter_list.append(batch)
+                        batch = [index]
+                        total_len = len_list[index]
+                    else:
+                        batch.append(index)
+                        total_len += len_list[index]
+        else:
+            for i in range(self.buckets[k].size(0)):
```
Comment (ejguan) [inline examples/hubert/dataset/hubert_dataset.py:100]: Please correct me if I'm wrong. Where does `k` come from in this clause?

**Review by ejguan — APPROVED** [review]
It looks good to me in terms of the implementation of sampler and dataset. Just having several questions about the internal logic of sampler. I won't block you because of my clarification questions.

File: examples/hubert/dataset/hubert_dataset.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+            dataset (str): The dataset for training. Options: [``librispeech``, ``librilight``].
+            subset (str): The subset of the dataset. Options: [``train``, ``valid``].
+
+        Returns:
+            (np.array): The numpy arrary that contains the labels for each audio file.
+        """
+        with open(label_dir / f"{dataset}_{subset}.pt") as f:
+            labels = [line.rstrip() for line in f]
+            labels = [[int(ele) for ele in labels[i].split()] for i in self.ind_list]
+        return np.asarray(labels, dtype=object)
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:223]: Okay, sounds good.

File: examples/hubert/dataset/hubert_dataset.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+
+    def __getitem__(self, index):
+        waveform = self._load_audio(index)
+        length = waveform.shape[1]
+        label = torch.Tensor(self.labels[index])
+        return (waveform, label, length)
+
+
+class collate_fn_hubert:
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:232]: You are right. It should be capital case since it's a class

File: examples/hubert/dataset/hubert_dataset.py (hunk 8 above)
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:100]: Oops, I didn't test the batch size option. I will fix it.

File: examples/hubert/dataset/hubert_dataset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ejguan) [reply examples/hubert/dataset/hubert_dataset.py:223]: How about `np.int`? IIRC, to avoid copy, the data structure needs to share a single ref_count. Like numpy array or Tensor, the whole data structure uses one ref_count. For python array, every single element within it is going to have a separate ref_count.
Basically, we want python only copy the pointer toward the data, rather than all data in multiprocessing.

File: examples/hubert/dataset/hubert_dataset.py (hunk 7 above)
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:89]: Good question. I tested by shuffling the bucket keys. If I choose a high ``bucket_size`` and ``crop`` option instead of ``padding``, it's very likely that the mini-batch will have different number of tokens. So I will keep the keys sorted here, but shuffle the indices inside each bucket.

File: examples/hubert/dataset/hubert_dataset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:223]: I see. I tried with ``np.int`` but it doesn't work if the lists have different lengths. Padding may work, but that may increase the memory usage.
Do you think ``np.empty`` with the max_length will help save the memory?

[the remark this answers is not in the document]
Comment (ejguan) [reply examples/hubert/dataset/hubert_dataset.py:225]: How about not splitting string here and using `np.string_`? -> `self.labels` is going to be list of string.
At the time that `__getitem__` is invoked, split each string to integer list and create a tensor from the list.

File: examples/hubert/dataset/hubert_dataset.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+        """
+        with open(label_dir / f"{dataset}_{subset}.pt") as f:
+            labels = [line.rstrip() for line in f]
+            labels = [[int(ele) for ele in labels[i].split()] for i in self.ind_list]
+        return np.asarray(labels, dtype=object)
+
+    def __getitem__(self, index):
+        waveform = self._load_audio(index)
+        length = waveform.shape[1]
+        label = torch.Tensor(self.labels[index])
```
Comment (ejguan) [inline examples/hubert/dataset/hubert_dataset.py:230]: BTW, pytorch prefers using `torch.tensor()` to create Tensor rather than `torch.Tensor`

**Review by ejguan — APPROVED** [review]
LGTM.

**Review by carolineechen — APPROVED** [review]
